src/esperanza/finalizationstate.cpp

```diff
@@ -178,19 +179,18 @@ void FinalizationState::IncrementDynasty() {
 }
 
 ufp64::ufp64_t FinalizationState::GetCollectiveRewardFactor() {
-  uint32_t epoch = m_current_epoch;
   bool isLive = GetEpochsSinceFinalization() <= 2;
 
   if (!DepositExists() || !isLive) {
     return 0;
   }
 
   ufp64::ufp64_t curVoteFraction = ufp64::div_2uint(
-      GetCheckpoint(epoch - 1).GetCurDynastyVotes(m_expected_source_epoch),
+      GetCheckpoint(m_current_epoch - 2).GetCurDynastyVotes(m_expected_source_epoch - 1),
```

Comment:
explanation of changes:
1. `m_current_epoch - 2`. We store `CurDynastyVotes` for the `target_epoch` which is always `-1` to the `m_current_epoch` (see: https://github.com/dtr-org/unit-e/blob/master/src/esperanza/finalizationstate.cpp#L500-L508) and an additional `-1` as this function is called inside of `InitialzeEpoch` where `m_current_epoch` is already incremented (see: https://github.com/dtr-org/unit-e/blob/master/src/esperanza/finalizationstate.cpp#L99-L101)
2. `m_expected_source_epoch - 1` is because we started to update source/target after we processed the checkpoint however it used to be at the end of `InitializeEpoch` so at the time of this call `m_expected_source_epoch` wasn't incremented and we didn't adjust this function accordingly. The rationale to update `source/target` when the checkpoint is processed is that the node can already include its vote for the 1st block otherwise if it decides to create the vote right in the first block, that vote will have incorrect source/target.